### DISCUSSION TITLE
How to refuse upgrade and handle the error client side ?

### DISCUSSION THREAD

**Discussion by Sceat** [uNetworking/uWebSockets.js#994] [Q&A] [answered]
I don't see a way to retrieve a cause on the client before the socket was opened.

### Server:

```js
App().ws('/*', {
  upgrade: (response, request, context) => {
    response.end('User limit reached')
    // response.writeStatus('400 Limit reached')
  },
})
```

### Client
![image](https://github.com/uNetworking/uWebSockets.js/assets/11330271/694083e0-69c9-4da7-9286-4530d5a0eda5)

**Comment by uNetworkingAB** [comment]
writeStatus must come before end

**Comment by e3dio** [comment, marked as answer]
It's standard spec that the websocket client does not display information about the connection open error. If you want to send info about the close to client you need to open connection then immediately close it with code/reason

**Comment by Sceat** [comment]
> writeStatus must come before end

I tried both

> It's standard spec that the websocket client does not display information about the connection open error. If you want to send info about the close to client you need to open connection then immediately close it with code/reason

Alright I see so we can't optimize that 👍🏻

**Comment by uNetworkingAB** [comment]
Ah. Yeah just let the connection open and immediately close it. That is not more costly, anyways and you get a standard way of delivering standard error codes and reason to the client.
